### ISSUE TITLE
scBubbletree

### ISSUE THREAD

**Issue by snaketron** [Bioconductor/Contributions#2781]
Update the following URL to point to the GitHub repository of
the package you wish to submit to _Bioconductor_

- Repository: https://github.com/snaketron/scBubbletree

Confirm the following by editing each check box to '[x]'

- [x] I understand that by submitting my package to _Bioconductor_,
  the package source and all review commentary are visible to the
  general public.

- [x] I have read the _Bioconductor_ [Package Submission][2]
  instructions. My package is consistent with the _Bioconductor_
  [Package Guidelines][1].

- [x] I understand Bioconductor [Package Naming Policy][9] and acknowledge
  Bioconductor may retain use of package name.

- [x] I understand that a minimum requirement for package acceptance 
  is to pass R CMD check and R CMD BiocCheck with no ERROR or WARNINGS. 
  Passing these checks does not result in automatic acceptance. The 
  package will then undergo a formal review and recommendations for 
  acceptance regarding other Bioconductor standards will be addressed.

- [x] My package addresses statistical or bioinformatic issues related
  to the analysis and comprehension of high throughput genomic data.

- [x] I am committed to the long-term maintenance of my package. This
  includes monitoring the [support site][3] for issues that users may
  have, subscribing to the [bioc-devel][4] mailing list to stay aware
  of developments in the _Bioconductor_ community, responding promptly
  to requests for updates from the Core team in response to changes in
  _R_ or underlying software.
  
- [x] I am familiar with the [Bioconductor code of conduct][7] and 
  agree to abide by it.

I am familiar with the essential aspects of _Bioconductor_ software
management, including:

- [x] The 'devel' branch for new packages and features.
- [x] The stable 'release' branch, made available every six
      months, for bug fixes.
- [x] _Bioconductor_ version control using [Git][5]
  (optionally [via GitHub][6]).

For questions/help about the submission process, including questions about
the output of the automatic reports generated by the SPB (Single Package
Builder), please use the #package-submission channel of our Community Slack.
Follow the link on the home page of the [Bioconductor website][8] to sign up.

[1]: https://contributions.bioconductor.org/
[2]: https://bioconductor.org/developers/package-submission/
[3]: https://support.bioconductor.org
[4]: https://stat.ethz.ch/mailman/listinfo/bioc-devel
[5]: http://bioconductor.org/developers/how-to/git/
[6]: http://bioconductor.org/developers/how-to/git/sync-existing-repositories/
[7]: https://bioconductor.org/about/code-of-conduct/
[8]: https://bioconductor.org/
[9]: https://bioconductor.org/developers/package-submission/#naming

**Comment by lshep** [comment]
@hpages  / @jwokaty  it looks like there is a system dependency that needs to be installed that doesn't necessarily come from this package but from a dependency of a dependency (Seurat utilizing leiden).  See [leiden gihub]( https://github.com/TomKellyGenetics/leiden)  that has some needed pip installs.  Please comment if this is acceptable to add to the builders.

**Comment by snaketron** [comment]
Any idea how to solve this error?

> * ERROR: Packages providing 1 object(s) used in this package should
>       be imported in the NAMESPACE file, otherwise packages importing
>       this package may fail.

No external package data is used.

Two datasets are included in dir /data: data_500 and data_ccl and both dataset are loaded in the .Rd
examples as data("d_500", package = "scBubbletree") and data("d_ccl", package = "scBubbletree").

**Comment by lshep** [comment]
If you run BiocCheck locally there will be a scBubbletree.BiocCheck with a more detailed 00BiocCheck.log.  This will give you more details about all of the ERROR/WARNING/NOTES .  The ERROR in question is detailed `scales::percent in get_cat_tiles()` is not in NAMESPACE.  Also the license in the DESCRIPTION should match the LICENSE file provided - the file provided is GPL-3.  Please correct these and other errors, warnings, and notes before @HelenaLC does a more thorough review.

**Comment by snaketron** [comment]
> If you run BiocCheck locally there will be a scBubbletree.BiocCheck with a more detailed 00BiocCheck.log. This will give you more details about all of the ERROR/WARNING/NOTES . The ERROR in question is detailed `scales::percent in get_cat_tiles()` is not in NAMESPACE. Also the license in the DESCRIPTION should match the LICENSE file provided - the file provided is GPL-3. Please correct these and other errors, warnings, and notes before @HelenaLC does a more thorough review.

Thanks for the info @lshep , previous errors+warnings are now fixed.

Two errors remain:
1. leidenalg software needs to be installed to solve this, as mentioned earlier by @lshep 
> get_r(...)` threw an error.
>   Message: Cannot find Leiden algorithm, please install through pip (e.g. pip install leidenalg).

2. I added the tag 'scBubbletree' to my profile yesterday, perhaps this error will be resolved later on?
> Maintainer must add package name to Watched Tags on the support site; 
> Edit your Support Site User Profile to add Watched Tags.

I will solve the remaining NOTEs in the next days.

**Comment by jwokaty** [comment]
I installed the stable version of `leiden` from CRAN on nebbiolo2. If this package is accepted, we'll install it on the other builders.

**Comment by hpages** [comment]
Based on the error reported by the SPB, It looks like **scBubbletree** requires _Python module_ `leidenalg`.

@snaketron Can you confirm this? If that's the case then you need to add a `SystemRequirements` field to your `DESCRIPTION` file that looks like this:
```
SystemRequirements: Python (>= 3), leidenalg
```
Note that you can also specify what minimum versions of Python 3 and/or `leidenalg` are required e.g. with something like `Python (>= 3.6), leidenalg (>= 0.21.2)` (I just made up versions here, please replace with appropriate versions).

Thanks,
H.

**Comment by hpages** [comment]
I should add that this requirement actually comes from the **leiden** package from CRAN, which is a hard dep of the **Seurat** package, which you depend on. FWIW other Bioconductor packages already depend directly or indirectly on **Seurat** e.g. [**CelliD**](https://bioconductor.org/packages/CelliD), [**phemd**](https://bioconductor.org/packages/phemd), [**scAlign**](https://bioconductor.org/packages/scAlign), [**scTreeViz**](https://bioconductor.org/packages/scTreeViz), [**Spaniel**](https://bioconductor.org/packages/Spaniel), + a dozen more...

The **leiden** package only exports one function, the `leiden()` function, and this function requires _Python module_ `leidenalg` to work. The fact that **leiden** doesn't list this hard requirement in `SystemRequirements` is wrong. And the fact that it gets away with it by disabling the examples in the `leiden()` man page (the only man page in the package!) is even more wrong. Unfortunately, that's what **reticulate**-based packages tend to do on CRAN i.e. they disable most or all their examples in order to pass `R CMD check`! :disappointed:

**Comment by snaketron** [comment]
> Based on the error reported by the SPB, It looks like **scBubbletree** requires _Python module_ `leidenalg`.
> 
> @snaketron Can you confirm this? 

@hpages 

Seurat's function FindClusters offers 4 clustering algorithms and Leiden is one of them. So yes, if the user decides to use Leiden then scBubbletree requires the python leidenalg module. 

I assume that if an R-package only in some instances requires some other software that it should nevertheless list this software as a hard requirement (would you agree?)


> If that's the case then you need to add a `SystemRequirements` field to your `DESCRIPTION` file that looks like this:
> 
> ```
> SystemRequirements: Python (>= 3), leidenalg
> ```
> Note that you can also specify what minimum versions of Python 3 and/or `leidenalg` are required e.g. with something like `Python (>= 3.6), leidenalg (>= 0.21.2)` (I just made up versions here, please replace with appropriate versions).

Thank you! I will add the appropriate field.

Thank you @jwokaty for installing leiden/leidenalg.

**Comment by hpages** [comment]
> I assume that if an R-package only in some instances requires some other software that it should nevertheless list this software as a hard requirement (would you agree?)

That's a great question. For R-package inter dependencies, we can choose between the `Depends`, `Imports`, `LinkingTo`, or `Suggests` field, to describe the type of dependencies. For example `Depends` and `Imports` are for _hard_ dependencies, and `Suggests` is for _soft_ dependencies.

But there's nothing like this for _external_ requirements. All the external requirements go in the `SystemRequirements` field, whether they are _hard_ or _soft_. In your case it seems that Python module `leidenalg` could be considered a _soft_ requirement, because you only need it for some specific use cases. However, your package won't pass `R CMD check` without `leidenalg` present on the machine, so, for that reason only, `leidenalg`  would need to be listed in the `SystemRequirements` field.

Hope this makes sense.

Thanks,
H.

**Comment by jwokaty** [comment]
Sorry, it errored because I didn't install the python package correctly; however, I installed it correctly and reran your build. There's still an error but it's related to a configuration on the support site for the maintainer.

**Comment by hpages** [comment]
The new error is probably on our side. I just checked and the **DelayedArray** package _is_  installed on nebbiolo2. Somehow it wasn't loadable when the SPB tried to build or check **scBubbletree** one hour ago. Could be bad timing between the SPB and BBS (the system running the Bioconductor daily builds) e.g. the latter was re-installing **DelayedArray** when the former was checking **scBubbletree**.

Please bump **scBubbletree**'s version again to trigger a new build. Third time is the charm? Fingers crossed :crossed_fingers:

**Comment by snaketron** [comment]
Will try tomorrow again :)

Many thanks to the Bioconductor Team for the support!

**Comment by hpages** [comment]
Well, this is just this BiocCheck error now so there's progress :wink: 
```
* Checking for bioc-devel mailing list subscription...
    * ERROR: Maintainer must subscribe to the bioc-devel mailing list.
      Subscribe here: https://stat.ethz.ch/mailman/listinfo/bioc-devel
```

**Comment by HelenaLC** [comment]
First round of comments below! Note that many are suggestions and not strictly required. Please comment back here with what has/not been addressed how/why not, or any comments regarding any needed clarification or questions you might have; happy to help/discuss. Cheers!

# DESCRIPTION

- [ ] Consider adding a `URL:` field pointing to the code repo, 
and a `BugReports:` field for reporting issues (e.g., to Github).

# Documentation

- [ ] (optional) We would highly encourage having a package man page so users 
would be directed to the main functions of the package with `?scBubbletree`.

# Vignette

- [ ] You have a neat description of the package's purpose. However, the vignette 
starts out with describing the example data and jumping right into its analysis. 
Perhaps you could include a brief abstract (in the YAML header) and/or Introduction 
section describing the package's motivation, perhaps summarizing key functionality, etc. 

- [ ] To direct users to external packages and highlight/make them distinguishable 
from general code, you can hyperlink them via `BiocStyle::CRANpkg("package_name")` 
(or `Githubpkg` or `Biocpkg`, depending on where the package is available from).

- [ ] Lines 558+: The formatting is off here as you are using single-back ticks 
to start quotation, which is interpreted as in-line code (see screenshot below). 
You can use plain single-/double-quotes instead, as these will be handled correctly 
when markdown is converted to LaTeX; for a discussion on this, see [here](https://tex.stackexchange.com/questions/431626/left-quotation-marks-when-using-rmarkdown-and-knitr).

<img width="714" alt="image" src="https://user-images.githubusercontent.com/14542264/193239372-fca7bcdd-8333-45ed-80d1-7ae9223b839a.png">

# Data

- [ ] The example data `d_ccl` and `d_500` are well described in their *man* pages, 
and some source information is provided as well. However, please also include the 
code that was used to generate these subsets in the `inst/script` directory; see 
[here](http://contributions.bioconductor.org/docs.html?q=inst/sc#doc-inst-script). 
You currently have this as commented out code in the vignette, which (at your own 
preference) could be removed if the code is provided in a separate script. 

# Code

- [ ] I found some functions rather hard to follow as they are cluttered with 
validity checks and messages; see also the related `BiocCheck` NOTE that 
"The recommended function length is 50 lines or less. There are 51 functions 
greater than 50 lines." You have already very nicely placed these checks in 
`check_input` functions. Would it not be possible to pull these out of the 
main functions, and place them either inside the same script but above/below, 
or in a separate `checks.R` script (or similar)? Of course, that would also 
require renaming to, e.g., `check_input_get_gini()` to avoid duplicates.

- [ ] Related to the above, I believe that some arguments are shared between 
functions, and hence include identical checks. E.g., `get_bubbletree_kmeans` 
lines 235+ is the same as `get_cat_tiles` lines 151+, checking the `round_digits` 
argument. Would it not be easier to define a `check_round_digits` function that 
is called in all relevant places?

- [ ] Doing a package-wide search for `message(`, I found that most scripts 
print information on progress. However, there is no way for users to suppress 
these at the moment without, e.g., using `suppressMessages()`. It would be nice 
to have affected functions offer a `verbose=TRUE/FALSE` argument that let's users 
control whether or not they want messages printed to the console. i.e., 

```r
foo <- function(..., verbose = TRUE) {
  ...
  if (verbose) 
    message("text1")
  ...
  if (verbose) 
    message("text1")
  ...
}
```

- [ ] `ggplot2`-related R CMD check NOTES "no visible binding for global variable ..." 
can be avoided by using `aes_string` instead of `aes`, which accepts a character string instead of symbolic. E.g., in `util.R`'s `get_dendrogram`, line 270:

```r
# line 239:
ggplot2::aes(size = c) # this gives a NOTE
ggplot2::aes_string(size = "c") # this doesn't

# line 270:
ggplot2::aes(label=label, subset=isTip==FALSE) # this gives a NOTE
ggplot2::aes_string(label="label", subset="isTip==FALSE") # this doesn't
```

- [ ] Somewhat related, if `ggplot` aesthetics are a more lengthy computation 
that is done on the fly, it's sometimes easier and cleaner to define a new 
(intermediate) variable instead, I think. E.g., line 252+:

```r
# this is quite confusing:
tree <- ggtree::ggtree(ph, ...)
...
tree <- tree + 
  geom_tiplab(mapping = ggplot2::aes(
    label=paste0(label, " (",
                 paste0(round(c/1000, digits = round_digits),
                        'K'), ', ', pct, "%)"))

# this is cleaner and perhaps easier to follow:
# ('sprintf' instead of 'paste(0)' might also help here!)
ph$.label <- sprintf("%s (%sK), %s%%", ph$label, 
  round(ph$c/1000, digits = round_digits), ph$pct)
tree <- tree + geom_tiplab(mapping = ggplot2::aes_string(label=".label")
```

# Tests

- [ ] Thanks for implementing unit tests! The current coverage is at 66%. 
Consider getting this closer to 100%; a few scripts are not tested at all. 

```r
# devtools::test_coverage()
scBubbletree Coverage: 66.03%
R/get_bubbletree_dummy.R: 0.00%
R/get_gini_k.R: 0.00%
R/get_num_tiles.R: 0.00%
R/get_num_violins.R: 0.00%
R/util.R: 77.86%
R/get_bubbletree_kmeans.R: 78.91%
R/get_bubbletree_louvain.R: 80.33%
R/get_cat_tiles.R: 86.36%
R/get_gini.R: 94.52%
R/get_k.R: 94.56%
R/get_r.R: 95.44%
```

# Other

- [ ] Please remember to update the `NEWS` file to track package changes. 
The package is currently at v0.99.10, and only v0.99.0 is currently listed.

- [ ] Please move `logo.png` to the `inst/extdata` directory to keep the package's base directory clean.

- [ ] The LICENCE is a template that contains lines such as "<one line to give the program's name 
and a brief idea of what it does.>" ... "Copyright (C) \<year\> \<name of author\>". 
Please go through the file, check for these, and fill them in.

**Comment by snaketron** [comment]
> First round of comments below! Note that many are suggestions and not strictly required. Please comment back here with what has/not been addressed how/why not, or any comments regarding any needed clarification or questions you might have; happy to help/discuss. Cheers!

Thank you for the review. Fixes have been pushed (0.99.11 and minor change in 0.99.12).


> 
> # DESCRIPTION
> * [ ]  Consider adding a `URL:` field pointing to the code repo,
>   and a `BugReports:` field for reporting issues (e.g., to Github).

Added URL and BugReports in Description

> 
> # Documentation
> * [ ]  (optional) We would highly encourage having a package man page so users
>   would be directed to the main functions of the package with `?scBubbletree`.

Added man/scBubbletree-package.Rd


> # Vignette
> * [ ]  You have a neat description of the package's purpose. However, the vignette
>   starts out with describing the example data and jumping right into its analysis.
>   Perhaps you could include a brief abstract (in the YAML header) and/or Introduction
>   section describing the package's motivation, perhaps summarizing key functionality, etc.

Background section added with some motivation.

Question: I have a few other vignettes (.Rmd reports and .html exports) which are too 
computationally heavy to be compiled regularly. Can I include the .htmls exports in 
/vignettes?   

> * [ ]  To direct users to external packages and highlight/make them distinguishable
>   from general code, you can hyperlink them via `BiocStyle::CRANpkg("package_name")`
>   (or `Githubpkg` or `Biocpkg`, depending on where the package is available from).

Added BiocStyle::CRAN/Biocpkg formatting in vignette

> * [ ]  Lines 558+: The formatting is off here as you are using single-back ticks
>   to start quotation, which is interpreted as in-line code (see screenshot below).
>   You can use plain single-/double-quotes instead, as these will be handled correctly
>   when markdown is converted to LaTeX; for a discussion on this, see [here](https://tex.stackexchange.com/questions/431626/left-quotation-marks-when-using-rmarkdown-and-knitr).


> <img alt="image" width="714" src="https://user-images.githubusercontent.com/14542264/193239372-fca7bcdd-8333-45ed-80d1-7ae9223b839a.png">

Now fixed


> # Data
> * [ ]  The example data `d_ccl` and `d_500` are well described in their _man_ pages,
>   and some source information is provided as well. However, please also include the
>   code that was used to generate these subsets in the `inst/script` directory; see
>   [here](http://contributions.bioconductor.org/docs.html?q=inst/sc#doc-inst-script).
>   You currently have this as commented out code in the vignette, which (at your own
>   preference) could be removed if the code is provided in a separate script.

Two R scripts (for each dataset) added in /inst/script

> # Code
> * [ ]  I found some functions rather hard to follow as they are cluttered with
>   validity checks and messages; see also the related `BiocCheck` NOTE that
>   "The recommended function length is 50 lines or less. There are 51 functions
>   greater than 50 lines." You have already very nicely placed these checks in
>   `check_input` functions. Would it not be possible to pull these out of the
>   main functions, and place them either inside the same script but above/below,
>   or in a separate `checks.R` script (or similar)? Of course, that would also
>   require renaming to, e.g., `check_input_get_gini()` to avoid duplicates.

Main changes:
* input_checks functions are now outside the main (exported) functions -> main functions are now considerably smaller
* general input_checks functions are now listed in a new script /R/input_checks.R (duplicate checks are thereby avoided, this addresses the point below)

> * [ ]  Related to the above, I believe that some arguments are shared between
>   functions, and hence include identical checks. E.g., `get_bubbletree_kmeans`
>   lines 235+ is the same as `get_cat_tiles` lines 151+, checking the `round_digits`
>   argument. Would it not be easier to define a `check_round_digits` function that
>   is called in all relevant places?

Fixed (see my earlier comment)

> * [ ]  Doing a package-wide search for `message(`, I found that most scripts
>   print information on progress. However, there is no way for users to suppress
>   these at the moment without, e.g., using `suppressMessages()`. It would be nice
>   to have affected functions offer a `verbose=TRUE/FALSE` argument that let's users
>   control whether or not they want messages printed to the console. i.e.,
> 
> ```r
> foo <- function(..., verbose = TRUE) {
>   ...
>   if (verbose) 
>     message("text1")
>   ...
>   if (verbose) 
>     message("text1")
>   ...
> }
> ```

Input parameter 'verbose' added


> * [ ]  `ggplot2`-related R CMD check NOTES "no visible binding for global variable ..."
>   can be avoided by using `aes_string` instead of `aes`, which accepts a character string instead of symbolic. E.g., in `util.R`'s `get_dendrogram`, line 270:
> 
> ```r
> # line 239:
> ggplot2::aes(size = c) # this gives a NOTE
> ggplot2::aes_string(size = "c") # this doesn't
> 
> # line 270:
> ggplot2::aes(label=label, subset=isTip==FALSE) # this gives a NOTE
> ggplot2::aes_string(label="label", subset="isTip==FALSE") # this doesn't
> ```

Fixed both by replacing aes with aes_string

> * [ ]  Somewhat related, if `ggplot` aesthetics are a more lengthy computation
>   that is done on the fly, it's sometimes easier and cleaner to define a new
>   (intermediate) variable instead, I think. E.g., line 252+:
> 
> ```r
> # this is quite confusing:
> tree <- ggtree::ggtree(ph, ...)
> ...
> tree <- tree + 
>   geom_tiplab(mapping = ggplot2::aes(
>     label=paste0(label, " (",
>                  paste0(round(c/1000, digits = round_digits),
>                         'K'), ', ', pct, "%)"))
> 
> # this is cleaner and perhaps easier to follow:
> # ('sprintf' instead of 'paste(0)' might also help here!)
> ph$.label <- sprintf("%s (%sK), %s%%", ph$label, 
>   round(ph$c/1000, digits = round_digits), ph$pct)
> tree <- tree + geom_tiplab(mapping = ggplot2::aes_string(label=".label")
> ```

Fixed as suggested.


> # Tests
> * [ ]  Thanks for implementing unit tests! The current coverage is at 66%.
>   Consider getting this closer to 100%; a few scripts are not tested at all.
> 
> ```r
> # devtools::test_coverage()
> scBubbletree Coverage: 66.03%
> R/get_bubbletree_dummy.R: 0.00%
> R/get_gini_k.R: 0.00%
> R/get_num_tiles.R: 0.00%
> R/get_num_violins.R: 0.00%
> R/util.R: 77.86%
> R/get_bubbletree_kmeans.R: 78.91%
> R/get_bubbletree_louvain.R: 80.33%
> R/get_cat_tiles.R: 86.36%
> R/get_gini.R: 94.52%
> R/get_k.R: 94.56%
> R/get_r.R: 95.44%
> ```

Tests added for all exported functions. The coverage now is close to 90%.

I will try to add more checks for the R/util.R in the following days.



> # Other
> * [ ]  Please remember to update the `NEWS` file to track package changes.
>   The package is currently at v0.99.10, and only v0.99.0 is currently listed.

More news added

> * [ ]  Please move `logo.png` to the `inst/extdata` directory to keep the package's base directory clean.

logo -> inst/extdata

> * [ ]  The LICENCE is a template that contains lines such as "<one line to give the program's name
>   and a brief idea of what it does.>" ... "Copyright (C) <year> <name of author>".
>   Please go through the file, check for these, and fill them in.

Added missing lines in the LICENCE file


@HelenaLC please let me know if something is missing/unclear.
Best

**Comment by HelenaLC** [comment]
Thanks for addressing these gaps / suggestions for improvements so quickly; looks great!! The key functions are indeed much more compact and easier to follow, and I'm getting ~93% coverage now with at least >80% lines covered per function. All other points have been resolved as well, though, yes, it would be nice to further populate the unit tests with future development - e.g., as issues arise and especially when implementing new features. I have only one remaining comment:

In `inst/script/get_d_ccl.R`, line 6's comment states "download the data from:" followed by a non-comment URL ("https://..."). This is not valid R code, and the script itself is not executable / reproducible as it is. Perhaps you could modify these lines as to actually download the data (e.g., into a temporary directory) and load it from there?

**Comment by snaketron** [comment]
> Thanks for addressing these gaps / suggestions for improvements so quickly; looks great!! The key functions are indeed much more compact and easier to follow, and I'm getting ~93% coverage now with at least >80% lines covered per function. All other points have been resolved as well, though, yes, it would be nice to further populate the unit tests with future development - e.g., as issues arise and especially when implementing new features.

Thank you for the fast and constructive review. More unit tests will follow.

> I have only one remaining comment: 
> In `inst/script/get_d_ccl.R`, line 6's comment states "download the data from:" followed by a non-comment URL ("https://..."). This is not valid R code, and the script itself is not executable / reproducible as it is. Perhaps you could modify these lines as to actually download the data (e.g., into a temporary directory) and load it from there?

Now the two scripts in inst/script/ should be executable.

**Comment by HelenaLC** [comment]
Again, thanks for the prompt fix. Beyond the initial stylistic / basic review, I found the following rather significant "issue" with `scBubbletree`'s current structure:

**major:**

- The `DESCRIPTION` lists `SummarizedExperiment` under `imports:`, however, the package isn't used anywhere (except in the context of `is(..., "SummarizedExperiment")`, which doesn't require it as a dependency).

- `Seurat` is not part of Bioconductor. While it is of course okay to support `SeuratObject`s as input, `scBubbletree` should also support Bioc's equivalent of single-cell data in order to be interoperable with existing Bioconductor infrastructure, namely, the `SingleCellExperiment` class. See [here](https://github.com/Bioconductor/Contributions/issues/2610) for another contribution where a similar discussion is ongoing.  

- Ideally, one would define functions for the `SingleCellExperiment` (since the package is aimed at being available through Bioc), and demonstrate in examples / the vignette how to run the method with `Seurat` using the `as.SingleCellExperiment` conversion function. That way, both frameworks would be supported. Alternatively, but methods for both inputs could be supported, however, at the cost of harder maintenance (e.g., keeping up with an ever-changing `Seurat`-infrastructure).

**minor:** (...but worthwhile considering!)

- Have you considered using `roxygen` for documentation as opposed to 1) writing `.man` pages from scratch and 2) using `package::function` throughout the package? Generally, I would argue that 1) `roxygen` is much easier / cleaner, 2) the documentation goes in the same place as the R code (easier to follow / update from a review / developer perspective), and 3) selectively importing via `@importFrom package function1 function2` omits the need for the rather verbose `package::function`-syntax (except in rare cases), and currently used functions will be listed explicitly in the `NAMESPACE` file. See [here](https://cran.r-project.org/web/packages/roxygen2/vignettes/roxygen2.html) for a quick overview.

**Comment by snaketron** [comment]
> Again, thanks for the prompt fix. Beyond the initial stylistic / basic review, I found the following rather significant "issue" with `scBubbletree`'s current structure:
> 
> **major:**
> 
> * The `DESCRIPTION` lists `SummarizedExperiment` under `imports:`, however, the package isn't used anywhere (except in the context of `is(..., "SummarizedExperiment")`, which doesn't require it as a dependency).

I will remove the SummarizedExperiment dependency and the lines of code where this is 
used. This is old code that should have already been removed.

> * `Seurat` is not part of Bioconductor. While it is of course okay to support `SeuratObject`s as input, `scBubbletree` should also support Bioc's equivalent of single-cell data in order to be interoperable with existing Bioconductor infrastructure, namely, the `SingleCellExperiment` class. See [here](https://github.com/Bioconductor/Contributions/issues/2610) for another contribution where a similar discussion is ongoing.
> * Ideally, one would define functions for the `SingleCellExperiment` (since the package is aimed at being available through Bioc), and demonstrate in examples / the vignette how to run the method with `Seurat` using the `as.SingleCellExperiment` conversion function. That way, both frameworks would be supported. Alternatively, but methods for both inputs could be supported, however, at the cost of harder maintenance (e.g., keeping up with an ever-changing `Seurat`-infrastructure).

My comments:
The main input of scBubbletree is a matrix (x) which represents low-dimensional projection (e.g. computed by PCA) of a scRNA-seq dataset. We know that SeuratObject (like SingleCellExperiment) store such projections as numeric matrices, and these matrices represent a small piece of the often much larger SingleCellExperiment or SeuratObject objects:

```
example(SingleCellExperiment, echo=FALSE)
x <- reducedDim(sce, "PCA") # num. matrix x is the main input of scBubbletree
```
or in Seurat

```
data(pbmc_small)
x <- Embeddings(object = pbmc_small[["pca"]]) # num. matrix x is the main input of scBubbletree
```

Hence, I have tried to avoid asking the user to provide as input a full SingleCellExperiment or SeuratObject. Else, scBubbletree would simply extract the projections and discard the rest of object's content. Given that popular packages such as Seurat keep on changing rapidly I thought that having a numeric matrix as input of scBubbletree will lead to higher usability.

Instead of adding interfaces for SingleCellExperiment or SeuratObject I would rather create another vignette demonstrating how to extract numeric projection matrices from these objects and providing them as input to scBubbletree. 

@HelenaLC  Would this work? Other suggestions?

I could implement an interface for the class LinearEmbeddingMatrix, but I do not see the point of doing this as SingleCellExperiment itself stores the projections as numeric matrices.

> **minor:** (...but worthwhile considering!)
> * Have you considered using `roxygen` for documentation as opposed to 1) writing `.man` pages from scratch and 2) using `package::function` throughout the package? Generally, I would argue that 1) `roxygen` is much easier / cleaner, 2) the documentation goes in the same place as the R code (easier to follow / update from a review / developer perspective), and 3) selectively importing via `@importFrom package function1 function2` omits the need for the rather verbose `package::function`-syntax (except in rare cases), and currently used functions will be listed explicitly in the `NAMESPACE` file. See [here](https://cran.r-project.org/web/packages/roxygen2/vignettes/roxygen2.html) for a quick overview.

I am familiar with the tool/functionality but have no practical experience with it, hence I have avoided it. Will have a look.

**Comment by HelenaLC** [comment]
Yes, absolutely. If a numeric matrix that represents a lower-dimensional cell embedding is indeed sufficient as input for `scBubbletree`, implementing methods to work with a matrix as input would be the simplest and most versatile approach. Then, you could demonstrate in the vignette/examples how to extract such a matrix from a `SeuratObject` (via `Embeddings()`) and `SingleCellExperiment` (via `reducedDim()`). Apologies if I missed this; I haven't looked into detail as to which components of the `SeuratObject` are being used. But if it's the reduced dimensions only, I think you could get away without depending on either `Seurat` nor `SingleCellExperiment`, thus being more flexible and greatly reducing the number of package dependencies.

**Comment by snaketron** [comment]
> Yes, absolutely. If a numeric matrix that represents a lower-dimensional cell embedding is indeed sufficient as input for `scBubbletree`, implementing methods to work with a matrix as input would be the simplest and most versatile approach. Then, you could demonstrate in the vignette/examples how to extract such a matrix from a `SeuratObject` (via `Embeddings()`) and `SingleCellExperiment` (via `reducedDim()`). Apologies if I missed this; I haven't looked into detail as to which components of the `SeuratObject` are being used. But if it's the reduced dimensions only, I think you could get away without depending on either `Seurat` nor `SingleCellExperiment`, thus being more flexible and greatly reducing the number of package dependencies.

Thanks for quick reply @HelenaLC !

Changes in v. 0.99.17 (in response to major points in the previous review phase):
* new (additional) vignette scBubbletree_inputs explaining main input A -> cell embeddings numeric matrix 
* removed unnecessary SummarizedExperiment dependency 

As soon as the paper is out on scBubbletree I expect more/continuous changes in 
accordance to what the "masses" want.

**Comment by HelenaLC** [comment]
Thanks! I'm happy to accept the package. However, I'd strongly encourage sticking the description of inputs into the main vignette, as it's sub-optimal from a user perspective to have key inputs described in a separate file. 
I also believe this can also be kept much shorter, because this is a very basic aspect of single-cell data, and both `Seurat` and `SingleCellExperiment` users will be familiar with accessing it. It is only relevant to mention that both type of workflows are supported as you are only reliant on `A`. 
For example, you don't need to print a lot of details on the input structure. I think it's sufficient to say something like "`A` should being a numeric matrix (with rows = observations and columns = features) that represents a low-dimensional cell embedding (e.g., PCA)". Then, just mention in the text (or have two short code chunks demonstrating) how to access `A` via `Embeddings()` (from `Seurat`) and `reducedDims()` (from `SingleCellExperiment`); and done. Overall, I think all this can be a single paragraph, and anything more lengthly doesn't really aid readability and usability of the vignette(s).

**Comment by snaketron** [comment]
Thank you, I will make the change.

**Comment by lshep** [comment]
The master branch of your GitHub repository has been added to Bioconductor's git repository.

To use the git.bioconductor.org repository, we need an 'ssh' key to associate with your github user name. If your GitHub account already has ssh public keys (https://github.com/snaketron.keys is not empty), then no further steps are required. Otherwise, do the following:

1. [Add an SSH key to your github account][]
1. [Submit your SSH key to Bioconductor][]

See further instructions at

https://bioconductor.org/developers/how-to/git/

for working with this repository. See especially

https://bioconductor.org/developers/how-to/git/new-package-workflow/
https://bioconductor.org/developers/how-to/git/sync-existing-repositories/

to keep your GitHub and Bioconductor repositories in sync.

Your package will be included in the next nigthly 'devel' build (check-out from git at about 6 pm Eastern; build completion around 2pm Eastern the next day) at

https://bioconductor.org/checkResults/

(Builds sometimes fail, so ensure that the date stamps on the main landing page are consistent with the addition of your package). Once the package builds successfully, you package will be available for download in the 'Devel' version of Bioconductor using `BiocManager::install("scBubbletree")`. The package 'landing page' will be created at

https://bioconductor.org/packages/scBubbletree

If you have any questions, please contact the bioc-devel mailing list (https://stat.ethz.ch/mailman/listinfo/bioc-devel); this issue will not be monitored further.

[Add an SSH key to your github account]: https://help.github.com/articles/connecting-to-github-with-ssh/
[Submit your SSH key to Bioconductor]: https://git.bioconductor.org/BiocCredentials/
